### PR TITLE
Example for simple Expr --> SQL conversion

### LINKED ISSUE #10524: Add an example of how to use the SQL parser/unparser API
Part of https://github.com/apache/datafusion/issues/9494

### Is your feature request related to a problem or challenge?

Having a good example helps to make features easier to use in DataFusion

In this case the usecase is programmatic construction of SQL using the DataFUsion `Expr` and `LogicalPlan` s using the  api from  https://github.com/apache/datafusion/issues/9726 

### Describe the solution you'd like

1. Add a new examples directory https://github.com/apache/arrow-datafusion/blob/main/datafusion-examples/examples.  Potential names `plan_to_sql.rs`
2. Add an example that shows how to create an expression using the expr_fn api (e.g. `col("x") > 5 AND col("b") = 'foo'`) and convert that to a string

### Describe alternatives you've considered

_No response_

### Additional context

I would expect to extend the same example to converting LogicalPlan --> SQL query as well once https://github.com/apache/datafusion/issues/8661 is sufficiently far along (it may be already, I just haven't checked)

### PR DESCRIPTION
**Pull request by edmondop**
## Which issue does this PR close?

Closes #10524 .

At the time of opening this PR, the example fails with:

```
thread 'main' panicked at datafusion-examples/examples/plan_to_sql.rs:43:5:
assertion `left == right` failed
  left: "a < Int32(5) OR a = Int32(8)"
 right: "a < 5 OR a = 8"
```

### CODE REVIEW

**Comment by edmondop** [comment]
Using the `expr_to_sql` api, we get the following error:

```
assertion `left == right` failed
  left: "((\"a\" < 5) OR (\"a\" = 8))"
 right: "a < 5 OR a = 8"
```

**Comment by alamb** [comment]
> Using the expr_to_sql api, we get the following error:


For anyone else following along, this conversation is happening on the ticket: https://github.com/apache/datafusion/issues/10524#issuecomment-2113100593

File: datafusion-examples/examples/plan_to_sql.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    // See how to evaluate expressions
+    simple_expr_to_sql_demo();
+
+    Ok(())
+}
+
+/// DataFusion can convert expressions to SQL
```
Comment (alamb) [inline datafusion-examples/examples/plan_to_sql.rs:39]: Love it!

**Review by alamb — COMMENTED** [review]
Thanks @edmondop  -- this is pretty sweet

It would also be good to add `plan_to_sql.rs` to the list of examples in https://github.com/apache/datafusion/tree/main/datafusion-examples#single-process

**Comment by edmondop** [comment]
> Thanks @edmondop -- this is pretty sweet
> 
> It would also be good to add `plan_to_sql.rs` to the list of examples in [`main`/datafusion-examples#single-process](https://github.com/apache/datafusion/tree/main/datafusion-examples?rgh-link-date=2024-05-15T19%3A06%3A13Z#single-process)

Examples are now more complete and readme is updated

File: datafusion-examples/README.md [hunk 2]
```diff
@@ -63,6 +63,7 @@ cargo run --example csv_sql
 - [`parquet_sql.rs`](examples/parquet_sql.rs): Build and run a query plan from a SQL statement against a local Parquet file
 - [`parquet_sql_multiple_files.rs`](examples/parquet_sql_multiple_files.rs): Build and run a query plan from a SQL statement against multiple local Parquet files
 - ['parquet_exec_visitor.rs'](examples/parquet_exec_visitor.rs): Extract statistics by visiting an ExecutionPlan after execution
+- - [`plan_to_sql.rs`](examples/plan_to_sql.rs): Generate SQL from Datafusion `Expr` and `LogicalPlan`
```
Comment (yyy1000) [inline datafusion-examples/README.md:66]: There seems an additional `-`?

**Review by alamb — APPROVED** [review]
Thank you @edmondop  -- this looks like a great start of these examples. I think @yyy1000's comments about the extra `-` should be addressed but it isn't required.

cc @devinjdangelo and @backkem (given you started this project)

**Comment by alamb** [comment]
I filed https://github.com/apache/datafusion/issues/10550 for hte logical plan version too

**Comment by backkem** [comment]
> Using the `expr_to_sql` api, we get the following error:
> 
> ```
> assertion `left == right` failed
>   left: "((\"a\" < 5) OR (\"a\" = 8))"
>  right: "a < 5 OR a = 8"
> ```

The current unparser is somewhat conservative for correctness sake; Always quoting identifiers and adding every set of parentheses possible. If we want to make the generated SQL more succinct, these steps will have to be made "smarter". We'll have to add in the math rules to avoid unneeded parentheses and (likely dialect specific) rules for determining of quoting is needed. Note that the latter likely involves listing out the reserved keywords for each dialect.

**Review by alamb — APPROVED** [review]
Thanks again @edmondop

**Comment by alamb** [comment]
> > Using the `expr_to_sql` api, we get the following error:
> > ```
> > assertion `left == right` failed
> >   left: "((\"a\" < 5) OR (\"a\" = 8))"
> >  right: "a < 5 OR a = 8"
> > ```
> 
> The current unparser is somewhat conservative for correctness sake; Always quoting identifiers and adding every set of parentheses possible. If we want to make the generated SQL more succinct, these steps will have to be made "smarter". We'll have to add in the math rules to avoid unneeded parentheses and (likely dialect specific) rules for determining of quoting is needed. Note that the latter likely involves listing out the reserved keywords for each dialect.

Thank you for that excellent explanation @backkem . Since this has now come up several times, I filed https://github.com/apache/datafusion/issues/10557 to track it

**Pull request merged**